### PR TITLE
Fix generic_matmul handling of size zero arrays

### PR DESCRIPTION
**Pull request by kshyatt**
Right now the `rmul!` *forces* `beta = 0` even when that should not be the case. However, the output should still be updated even if the input array is size 0.

### CODE REVIEW

File: test/hip_rocarray/blas.jl [hunk 1]
```diff
@@ -786,6 +786,22 @@ end
             end
         end
     end
+end
 
+@testset "Handling size zero arrays correctly" begin
+    for T in (Float32, Float64, ComplexF32, ComplexF64)
+        m, k, n = (12, 6, 0)
```
Comment (simeonschaub) [inline test/hip_rocarray/blas.jl]: Shouldn't it be k = 0 instead of n = 0 here? Otherwise, C0 is just empty and this doesn't make a difference, right?

**Review by simeonschaub — COMMENTED** [review]
Thanks! Could you add a test for the case where both `C0` and `A0` are vectors?

Maybe we can also add a test where `C0` is filled with `NaN` and beta is zero to make sure beta is a strong zero? I assume that's why you used `_rmul_or_fill!`

File: test/hip_rocarray/blas.jl (hunk 1 above)
Comment (kshyatt) [reply test/hip_rocarray/blas.jl]: Ah yes good catch

File: test/hip_rocarray/blas.jl [hunk 2]
```diff
@@ -786,6 +786,46 @@ end
...
+            Ccpu = copy(C0)
+            mul!(Ccpu, A0, B0, α, β)
+            Cgpu = ROCArray(copy(C0))
+            mul!(Cgpu, ROCArray(A0), ROCArray(B0), α, β)
+            @test Ccpu ≈ Array(Cgpu)
+        end
+        # matrix * vector 
+        C0 = rand(T, m)
+        A0 = rand(T, k)
+        B0 = rand(T, k, m)
```
Comment (simeonschaub) [inline test/hip_rocarray/blas.jl]: Should this be
```suggestion
        A0 = rand(T, m, k)
        B0 = rand(T, k)
```
?

File: test/hip_rocarray/blas.jl [hunk 3]
```diff
@@ -786,6 +786,46 @@ end
...
+        for (α, β) in ( (-one(T), one(T)), (one(T), 2*one(T)), (one(T), zero(T))) 
+            Ccpu = copy(C0)
+            mul!(Ccpu, A0, B0, α, β)
+            Cgpu = ROCArray(copy(C0))
+            mul!(Cgpu, ROCArray(A0), ROCArray(B0), α, β)
+            @test Ccpu ≈ Array(Cgpu)
+        end
+        # C0 is NaN and beta is zero to test beta is applied
+        C0 = rand(T, m, n)
+        C0[m-1, n-1] = T(NaN)
```
Comment (simeonschaub) [inline test/hip_rocarray/blas.jl]:
```suggestion
        C0 = fill(T(NaN), m, n)
```

Comment (kshyatt) [reply test/hip_rocarray/blas.jl]: Isn't C0 still CPU side at this point?

File: test/hip_rocarray/blas.jl (hunk 2 above)
Comment (kshyatt) [reply test/hip_rocarray/blas.jl]: You asked above for the case where C0 and A0 are vectors?

Comment (simeonschaub) [reply test/hip_rocarray/blas.jl]: Ah, sorry, that's my bad! I meant C0 and B0 in order to cover the gemv method

File: test/hip_rocarray/blas.jl (hunk 3 above)
Comment (simeonschaub) [reply test/hip_rocarray/blas.jl]: Oops, yes! I updated my suggestion

**Review by simeonschaub — APPROVED** [review]
Looks great, thanks!

**Review by luraess — APPROVED** [review]

**Comment by kshyatt** [comment]
Hmmm this passed for me locally...

**Comment by luraess** [comment]
Thanks!
